File: LaserPolishingModel/Util/Ripples.cs
using MathNet.Numerics.LinearAlgebra;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using static LaserPolishingModel.Util.MathNetExtensions;
using static LaserPolishingModel.Util.IEnumerableExtensions;

namespace LaserPolishingModel.Util
{
    public static class Ripples
    {
        const double GRIDSET = 0.2;

        public static Matrix<double> GenerateRipples(Vector<double> xVec, Vector<double> yVec, double afs, double d, double spot, double xStep, double yStep)
        {
            var r_spot = spot / 2;
            var ifs_inner = afs;
            var r_melt = d / 2;
            var r_peak = r_melt - r_spot;

            var ifs_outer = (ifs_inner * Math.Pow(r_peak, 3)) / (Math.Pow(r_peak, 3) - Math.Pow(r_melt, 3));
            var z_valley = -ifs_outer * (r_melt - r_peak);
            var z_peak = z_valley - (ifs_inner * r_peak);

            // generate single spot
            var spot_row_size = (int) Math.Ceiling(2 * r_melt / GRIDSET);
            var spot_col_size = spot_row_size;
            Matrix<double> spotShape = Matrix<double>.Build.Dense(spot_row_size, spot_col_size);
            Matrix<double> circularRegion = Matrix<double>.Build.Dense(spot_row_size, spot_col_size);

            var outer_region_offset = z_peak - ifs_outer * r_peak;
            Parallel.For(0, spot_row_size * spot_col_size, (i) =>
            {
                var row = i / spot_row_size;
                var col = i % spot_row_size;

                var x = (r_melt / xVec.Count()) * col - r_melt;
                var y = (-r_melt / yVec.Count()) * row + r_melt;

                var location = Math.Pow(x, 2) + Math.Pow(y, 2);

                if (location < Math.Pow(r_peak, 2))
                    spotShape[row, col] = ifs_inner * Math.Sqrt(location) + z_valley;
                else
                    spotShape[row, col] = ifs_outer * Math.Sqrt(location) + outer_region_offset;

                if (location < Math.Pow(r_peak + GRIDSET, 2))
                    circularRegion[row, col] = spotShape[row, col];
            });

            // overlap area & # of spots
            var roit_proj_area = Math.PI * Math.Pow(r_melt, 2);
            var x_area_overlap = AreaOverlap(r_melt, xStep);
            var y_area_overlap = AreaOverlap(r_melt, yStep);

            var x_overlap_pct = x_area_overlap / roit_proj_area;
            var y_overlap_pct = y_area_overlap / roit_proj_area;

            var x_spots = (int) (2 * Math.Ceiling(1 / (1 - x_overlap_pct)));
            var y_spots = (int) (2 * Math.Ceiling(1 / (1 - y_overlap_pct)));

            var grid_x_limits = Range(-r_melt, (x_spots - 1) * xStep + r_melt, GRIDSET);
            var grid_y_limits = Range(-r_melt, (y_spots - 1) * yStep + r_melt, GRIDSET);
            //var grid_coordinates = meshgrid(grid_x_limits, grid_y_limits);

            // create large grid
            Matrix<double> overlay_grid = Matrix<double>.Build.Dense(grid_y_limits.Count(), grid_x_limits.Count());

            // lay spots down in a snake-like fashion
            // i.e. (1, 1) -> (1, n) -> (2, n) -> (2, 1) -> (3, 1) ...

            var forward_y_list = Range(0, y_spots, 1);
            var backward_y_list = Range(y_spots - 1, 0, -1);

            for (int i = 0; i < x_spots; i++)
            {
                foreach (var j in (i % 2 == 1) ? forward_y_list : backward_y_list)
                {
                    var x_center_index = (int) (((i * xStep) + r_melt) / GRIDSET); // this gets us the true coordinate. 
                    var y_center_index = (int) (((j * yStep) + r_melt) / GRIDSET);
                    var x_start = x_center_index - (spot_col_size / 2);
                    var x_end = x_center_index + (spot_col_size / 2);
                    var y_start = y_center_index - (spot_row_size / 2);
                    var y_end = y_center_index + (spot_row_size / 2);

                    Parallel.For(x_start, x_end + 1, (x_index) =>
                    {
                        for (int y_index = y_start; y_index <= y_end; y_index++)
                        {
                            var location = Math.Pow(x_index - x_center_index, 2) + Math.Pow(y_index - y_center_index, 2);
                            if (location < Math.Pow((r_peak / GRIDSET) + 1, 2))
                                overlay_grid[y_index, x_index] = spotShape[y_index - y_start, x_index - x_start];
                        }
                    });
                }
            }

            // grab the lattice
            var unit_cell_start_row = (int) (r_melt / GRIDSET);
            var unit_cell_start_col = (int)(r_melt / GRIDSET);
            var unit_cell_num_rows = (int)(((x_spots - 1) * xStep - r_melt) / GRIDSET) - unit_cell_start_row;
            var unit_cell_num_cols = (int)(((y_spots - 1) * yStep - r_melt) / GRIDSET) - unit_cell_start_col;

            var unit_cell = overlay_grid.SubMatrix(unit_cell_start_row, unit_cell_num_rows, unit_cell_start_col, unit_cell_num_cols);
            unit_cell -= overlay_grid.RowSums().Sum() / (overlay_grid.ColumnCount * overlay_grid.RowCount);

            // now replicate the unit cell to a size close to the input dimension
            int num_cols = (int) (2 * Math.Ceiling((xVec.Count + 0.0) / unit_cell.ColumnCount));
            int num_rows = (int) (2 * Math.Ceiling((yVec.Count + 0.0) / unit_cell.RowCount));
            var lattice = unit_cell.repmat(num_rows, num_cols);
            
             



            return overlay_grid;
        }

        static double AreaOverlap(double radius, double step)
        {
            return 2 * Math.Pow(radius, 2) * Math.Acos(step / (2 * radius)) - (step / 2) * Math.Sqrt(4 * Math.Pow(radius, 2) - Math.Pow(step, 2));
        }

        /*
    %% Unit cell to lattice
    
    % now take the unit cell and replicate to a size close to XI * YI
    % once again, 2x factor is for fudge factor to allow interp2 to do a
    % better job
    lattice = repmat(unit_cell, 2 * ceil(size(XI, 2) / size(unit_cell, 2)), 2 *ceil(size(YI, 1) / size(unit_cell, 1)));
    lattice_x = 0:gridset:gridset * (size(lattice, 2) - 1);
    lattice_y = 0:gridset:gridset * (size(lattice, 1) - 1);
    lattice_y = lattice_y';
    
    %% Convert to Experimental Mesh
    lattice_x = lattice_x / 1000;
    lattice_y = lattice_y / 1000;
    
    ZI = interp2(lattice_x, lattice_y, lattice, XI, YI);*/
    }
}

File: backend/model/ripples.py
import matplotlib.pyplot as plt
import numpy as np
from scipy import optimize, interpolate
from .zygo import ZygoAsciiFile

class SpotParameters:
    """ 
    The SpotParameters class generates a single laser spot from the following parameters.
    """
    def __init__(self, ifs_inner, ifs_outer, r_melt, 
        r_peak, z_valley, z_peak, gridset):
        """ Constructs a new SpotParameters object with the specified parameters."""
        self.ifs_inner = ifs_inner
        self.ifs_outer = ifs_outer
        self.r_melt = r_melt
        self.r_peak = r_peak
        self.z_valley = z_valley
        self.z_peak = z_peak
        self.gridset = gridset

    def create_spot(self):
        """Generates the conical spot shape, and returns a tuple of the resultant surface. """
        num_pts = int((2 * self.r_melt) / self.gridset) + 1
        r_peak_sq = np.power(self.r_peak, 2)
        outer_region_offset = self.z_peak - self.ifs_outer * self.r_peak

        xy_bounds = np.linspace(-self.r_melt, self.r_melt, num_pts)
        x, y = np.meshgrid(xy_bounds, xy_bounds)

        grid = np.power(x, 2) + np.power(y, 2)
        spot = np.piecewise(
            grid, 
            [grid < r_peak_sq, grid >= r_peak_sq], 
            [lambda s: self.ifs_inner * np.sqrt(s) + self.z_valley, lambda s: self.ifs_outer * np.sqrt(s) + outer_region_offset]
        )

        return [x, y, spot]

def compute_zero_volume_parameters(ifs_inner, ifs_outer, r_melt, r_peak, z_valley, z_peak):
    """
    Computes the necessary parameters to have a mass-conserved spot. 3 of 6 parametes supplied
    must be falsey (empty list, false, empty strings, etc), otherwise we cannot run the 
    computation as we would be overconstrained.
    """

    def zero_vol_constraints(x):
        """ Defines the 6 constraints necessary to have a mass-conserving spot."""
        # [ifs_inner, ifs_outer, r_melt, r_peak, z_valley, z_peak]
        constraints = [
            x[1] - ((x[0] * (x[3] ** 3)) / (x[3] ** 3 - x[2] ** 3)),
            x[4] + x[1] * (x[2] - x[3]),
            x[5] - x[4] + (x[0] * x[3])
        ]

        if ifs_inner:
            constraints.append(x[0] - ifs_inner)
    
        if ifs_outer:
            constraints.append(x[1] - ifs_outer)

        if r_melt:
            constraints.append(x[2] - r_melt)

        if r_peak:
            constraints.append(x[3] - r_peak)

        if z_valley:
            constraints.append(x[4] - z_valley)

        if z_peak:
            constraints.append(x[5] - z_peak)

        return constraints

    # solve for the constraints, and return the solution
    sol = optimize.root(zero_vol_constraints, [ifs_inner, ifs_outer, r_melt, r_peak, z_valley, z_peak], method='lm')
    return sol.x

def area_overlap(radius, step):
    """ Determines the ratio of overlap between circles with the same radii with the specified center offset."""
    return 2 * (radius ** 2) * np.arccos(step / (2 * radius)) - (step / 2) * np.sqrt(4 * (radius ** 2) - (step ** 2))

def ripples(XI, YI, afs_inner, d, spot, step):
    """ Generates the ripple surface by overlaying multiple spots."""
    r_spot = spot / 2
    ifs_outer = 0
    ifs_inner = afs_inner
    r_melt = d / 2
    r_peak = r_melt - r_spot
    z_valley = 0
    z_peak = 0
    gridset = 0.2

    # first, compute the mass-conserving spot.
    zero_vol_params = compute_zero_volume_parameters(ifs_inner, ifs_outer, r_melt, r_peak, z_valley, z_peak)
    sp = SpotParameters(*zero_vol_params, gridset)
    x, y, z = sp.create_spot()

    # let step either be a scalar (so x and y offsets are the same)
    # or a vector [x, y]
    if isinstance(spot, float) or isinstance(spot, int):
        x_step = step
        y_step = step
    else:
        x_step = step[1]
        y_step = step[2]

    roit = (x ** 2 + y ** 2) < (r_melt + gridset * 0.5) ** 2
    circular_region = z[roit]
    
    # plt.figure()
    # surf(x, y, z, 'LineStyle', 'none')
    
    ## Now overlay those spots over a grid
    # compute overlay
    # use the overlap area as a measure of how many spots we need to
    # generate the unit cell
    roit_proj_area = np.pi * (r_melt ** 2)
    x_area_overlap = area_overlap(r_melt, x_step)
    y_area_overlap = area_overlap(r_melt, y_step)

    x_overlap_pct = x_area_overlap / roit_proj_area
    y_overlap_pct = y_area_overlap / roit_proj_area
  
    # note the 2x factor is a "fudge" factor - unit cells seem better
    # this way
    x_spots = int(2 * np.ceil(1 / (1 - x_overlap_pct)))
    y_spots = int(2 * np.ceil(1 / (1 - y_overlap_pct)))

    # compute the grid x/y bounds
    lower_left_coord = (-r_melt, -r_melt)
    upper_right_coord = ((x_spots - 1) * x_step + r_melt, (y_spots - 1) * y_step + r_melt)
    grid_num_x = int(np.ceil(upper_right_coord[0] - lower_left_coord[0]) / gridset) + 1
    grid_num_y = int(np.ceil(upper_right_coord[1] - lower_left_coord[1]) / gridset) + 1
    grid_x_limits = np.linspace(lower_left_coord[0], upper_right_coord[0], grid_num_x)
    grid_y_limits = np.linspace(lower_left_coord[0], upper_right_coord[0], grid_num_y)
    
    [overlay_x, overlay_y] = np.meshgrid(grid_x_limits, grid_y_limits)
    overlay_grid = np.zeros(np.shape(overlay_x))

    # go in a snake pattern in the y direction
    # i.e. (1, 1) -> (1, n) -> (2, n) -> (2, 1) -> (3, 1) ...
    for i in range(x_spots):
        y_list = range(y_spots)
        if i % 2 == 0:
            y_list = y_list[::-1]

        for j in y_list:
            x_center = i * x_step
            y_center = j * y_step
            circular_region_selector = ((overlay_x - x_center) ** 2 + \
                (overlay_y - y_center) ** 2) < (r_melt + gridset * 0.5) ** 2
            overlay_grid[circular_region_selector] = circular_region

    # figure
    # surf(overlay_x, overlay_y, overlay_grid, 'LineStyle', 'none');

    # Unit cell to lattice
    # grab 'unit' from overlayed_grid
    uc_x_lim = np.array([r_melt, (x_spots - 1) * x_step - r_melt])
    uc_y_lim = np.array([r_melt, (y_spots - 1) * y_step - r_melt])
    uc_selector = np.logical_and(
        np.logical_and(overlay_x > uc_x_lim[0], overlay_x <= uc_x_lim[1]),
        np.logical_and(overlay_y > uc_y_lim[0], overlay_y <= uc_y_lim[1]))
    uc = overlay_grid[uc_selector]

    # pull the unit cell out and normalize it
    unit_cell_x = uc_x_lim / gridset
    unit_cell_y = uc_y_lim / gridset
    unit_cell = np.reshape(uc, [int(unit_cell_x[1] - unit_cell_x[0]), int(unit_cell_y[1] - unit_cell_y[0])])
    unit_cell = unit_cell - np.mean(unit_cell)
    
    # now take the unit cell and replicate to a size close to XI * YI
    # once again, 2x factor is for fudge factor to allow interp2 to do a
    # better job
    x_repl = int(2 * np.ceil(np.size(XI) / np.shape(unit_cell)[1]))
    y_repl = int(2 * np.ceil(np.size(YI) / np.shape(unit_cell)[0]))
    lattice = np.matlib.repmat(unit_cell, x_repl, y_repl)
    lattice_x = np.linspace(0, gridset * (np.size(lattice, 1) - 1), np.size(lattice, 1))
    lattice_y = np.linspace(0, gridset * (np.size(lattice, 0) - 1), np.size(lattice, 0))
    lattice_y = lattice_y.T
    
    # Convert to Experimental Mesh through interpolation
    lattice_x = lattice_x / 1000
    lattice_y = lattice_y / 1000
    
    ip = interpolate.interp2d(lattice_x, lattice_y, lattice)
    return ip(XI, YI)
